[flang][OpenMP] Improve reduction of Scalar ArrayElement types

Stylie777 opened the pull request:
Currently, Flang does not correctly lower ArrayElement's when processing an OpenMP Reduction Clause correctly. Rather than lowering the array element, the whole array will be lowered. This leads to slower performance for the end user in their program.

This patch works to rectify this by rewriteing the parse tree while processing semantics. The use of an ArrayElement in an OpenMP Reduction Clause will be identified, and replaced with a temporary both in the reduction clause, and anywhere that array element is used within the respective DoConstruct. Once the DoConstruct has finished, if the ArrayElement has been used within the Do loop, the value of the temporary will be re-assigned to the array element. One limitation of this approach is that if the ArrayElement is not used, there is no available element in the parse tree to use to reassign the value, so its only done if used.

The reason for making the change in the parse tree is due to how ArrayElements are lowered. When lowering, the expression of the ArrayElement being used in the reduction is being substitued with the reference to the symbol. In this case, that would be the whole array. By replacing it with a temporary, it removes the issue of lowering a full array as it will be referencing the temporary instead. To address this in lowering would require a major rethink on how a considerable amount of non-OpenMP code is lowered and as such, not deemed the appropriate course of action for this specific case.

This process is done after the initial Semantics Pass as to not affect the checking of users original code. If the array element has been replaced, the first pass of semantics will need to be rerun to ensure all TypedExpr's are correctly captured otherwise the lowering will not function correctly. This step is only done if an ArrayElement is replaced.

Testing is covered by reduction17.f90. This checks both the parse tree, unparsing and HLFIR to ensure the temproary is being used in the reduction clause and Do loop. Assignment to, and reassignment from the ArrayElement and the Temporary is also considered to ensure this is inserted at the correct location.

reduction09.f90 has also been reformatted to rely on FileCheck. As the Parse Tree is changing, the output is different to that of the user, so we can no longer rely on test_symbols.py for this test. The same information is being checked, with test cases that cover using an ArrayElement in the Do loop, and not using the ArrayElement being covered.

Array Sections are not affected by this change, only uses of single array elements.

Stylie777 committed:
[flang][OpenMP] Fix reduction of Scalar ArrayElement types

Currently, Flang does not correctly lower ArrayElement's
when processing an OpenMP Reduction Clause correctly. Rather
than lowering the array element, the whole array will be lowered.
This leads to slower performance for the end user in their program.

This patch works to rectify this by rewriteing the parse tree
while processing semantics. The use of an ArrayElement in an
OpenMP Reduction Clause will be identified, and replaced with a
temporary both in the reduction clause, and anywhere that array
element is used within the respective DoConstruct. Once the
DoConstruct has finished, if the ArrayElement has been used within
the Do loop, the value of the temporary will be re-assigned to
the array element. One limitation of this approach is that if the
ArrayElement is not used, there is no available element in the parse
tree to use to reassign the value, so its only done if used.

The reason for making the change in the parse tree is due to how
ArrayElements are lowered. When lowering, the expression of the
ArrayElement being used in the reduction is being substitued with
the reference to the symbol. In this case, that would be the whole
array. By replacing it with a temporary, it removes the issue
of lowering a full array as it will be referencing the
temporary instead. To address this in lowering would require a
major rethink on how a considerable amount of non-OpenMP code
is lowered and as such, not deemed the appropriate course of
action for this specific case.

This process is done after the initial Semantics Pass as to
not affect the checking of users original code. If the array
element has been replaced, the first pass of semantics will
need to be rerun to ensure all TypedExpr's are correctly
captured otherwise the lowering will not function correctly.
This step is only done if an ArrayElement is replaced.

Testing is covered by reduction17.f90. This checks both the
parse tree, unparsing and HLFIR to ensure the temproary is being used in
the reduction clause and Do loop. Assignment to, and reassignment from
the ArrayElement and the Temporary is also considered to ensure this is
inserted at the correct location.

reduction09.f90 has also been reformatted to rely on FileCheck.
As the Parse Tree is changing, the output is different to that of
the user, so we can no longer rely on test_symbols.py for this test.
The same information is being checked, with test cases that cover using
an ArrayElement in the Do loop, and not using the ArrayElement being
covered.

Array Sections are not affected by this change, only uses of
single array elements.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +572,454 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+                          arrayElement.value().base.u);
+                      temp.setOriginalSubscriptInt(
+                          arrayElement.value().subscripts);
+                      return !temp.isSectionTriplet() ? true : false;
```
The "!temp.isSectionTriplet()" alone is already a bool.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +572,454 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+              bool rewritten{false};
+              for (auto object{objectList.begin()}; object != objectList.end();
+                  ++object) {
+                rewritten = std::visit(
```
This will only retain the last value after the loop terminates.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +572,454 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+void ReplacementTemp::createTempSymbol(
+    SourceName &source, Scope &scope, SemanticsContext &context) {
+  replacementTempSymbol_ =
+      const_cast<semantics::Scope &>(originalName_.symbol->owner())
```
No need for "semantics::".  This appears in a few places in your code.

kparzysz reviewed (commented):
Some initial comments.

The large indentations in some places make the code harder to read. Could you extract some of the visitor lambdas into separate entities?  e.g.
```
std::visit(
  [&](...) {
    [&] (...) {
    }
  }
)
```
to
```
auto foo = [&] (...) {
};
auto bar = [&] (...) {
  foo
};
std::visit(...bar...);
```
Or something like that to reduce the amount of leading whitespace.

kparzysz commented:
> One limitation of this approach is that if the ArrayElement is not used, there is no available element in the parse tree to use to reassign the value, so its only done if used.

Correct me if I'm reading the code wrong, but it seems like you do the reassignment back to the original array only when that element is used, but you do the replacement with a scalar in all cases.  If so, that will produce incorrect code:
```
subroutine foo(a, b)
  integer, intent(inout) :: a(10), b(10)
  integer :: i
  !$omp parallel do reduction(+: a(3))
  do i = 1, 10
    a(3) = a(3) + b(i)  ! update visible to callers of foo
  end do
  !$omp end parallel do
end
```

Stylie777 commented:
@kiranchandramohan From my understanding addressing this in Lowering is not possible. An array element is an expression and the reference to the Symbol is used, which in this case would be the full array. This is ok for Array Sections, but if the user is only using 1 single array element, it does not make sense to reduce the full array, hence the rewrite to the temp so it can pick up a symbol which is just a single value.

With this rewrite, the MLIR example you provided with a single integer being reduced is what ends up being generated, but its not possible to do this using the array element itself because of the process I explained above.

@kparzysz an example of the lowering as it is would be, say for a small program
```
program test
    integer :: a(2)
    integer :: i

    !$omp do reduction(+:a(2))
    do i=1,5
        a(2) = a(2) + i
    end do
    !$omp end do

end program test
```
When the reduction appears in High Level FIR, its generated as
```
reduction(byref @add_reduction_byref_box_2xi32 %7 -> %arg1 : !fir.ref<!fir.box<!fir.array<2xi32>>>)
```
As the user only needs 1 element of the array, it does not make sense to use the whole array in the reduction. This patch would then turn that into the following, with the assignment to and from the temp in the relevant places:
```
reduction(@add_reduction_i32 %7#0 -> %arg1 : !fir.ref<i32>) {
```

With the example you have given in the subroutine `foo(a, b)`, that does reassign after the loop correctly as it will detect the use of the array element within the loop and use one of those occurrences on ArrayElement in the Parse Tree to build the reassignment Expression. The issue I have is because I am moving the use of an ArrayElement for the reassignment Expression, if it's not used within the do loop, I cannot make it. But then if it's not used within the loop, do we need to reassign it back? Probably not which was the justification for taking this approach because in reality the value would not change anyway if you reduced the whole array.

kparzysz commented:
I'm ok with this approach, but I think we need to develop a solution in lowering in the longer term.

I'm concerned that doing something like the example above could potentially introduce subtle race conditions
```
tmp = a(10)
!$omp parallel reduction(+: tmp)
...
tmp = tmp + 10 
!$omp end parallel  ! a(10) would normally be up-to-date here on exit from the parallel region
a(10) = tmp         ! but it isn't until here
```

I think we can probably get away with this for now.

Stylie777 committed:
Code Style improvements & respond to comments

collate all std::visit's into lammbda functions to make them easier
to read.

Stylie777 reviewed (commented):
@kparzysz thanks for the initial review. I have also gone through and extracted the `std::visit` parts of the code into lambda functions so it is easier to follow and review as you suggested.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +572,468 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+  parser::OmpBeginLoopDirective &ompBeginLoop{
+      std::get<parser::OmpBeginLoopDirective>(ompLoop.t)};
+  std::list<parser::OmpClause> &clauseList{
+      std::get<std::optional<parser::OmpClauseList>>(ompBeginLoop.t)->v};
```
In case you didn't know, OpenMPLoopConstruct has a function "BeginDir" that returns a reference to the contained "OmpBeginLoopDirective".  The OmpBeginLoopDirective inherits from OmpDirectiveSpecification that has member function Clauses() that returns the list of clauses.
You could do
```
auto &clauseList{const_cast<std::list<parser::OmpClause>>(ompLoop).BeginDir().Clauses().v};
```
or
```
auto &clauseList{const_cast<parser::OmpClauseList>(ompLoop).BeginDir().Clauses()};
```
and use clauseList.v.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +572,468 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+  auto processReductionClause =
+      [&](parser::OmpClause::Reduction &clause) -> bool {
+    std::list<parser::OmpObject> &objectList =
+        std::get<parser::OmpObjectList>(clause.v.t).v;
```
FYI, there is a function in `flang/include/flang/Parser/openmp-utils.h` that gets OmpObjectList from any clause (or returns nullptr if the clause doesn't have one):
`const OmpObjectList *GetOmpObjectList(const OmpClause &clause);`

Stylie777 replied:
Done

kparzysz reviewed (commented):
Thanks, it looks a lot clearer now.

This code will only handle one reduction variable per reduction clause. For example, given
```
subroutine f00(x, y)
  integer :: x(:), y(:)
  integer :: i
  !$omp parallel do reduction(+: x(2), y(3))
  do i = 1, 200
    x(2) = x(2) + i
    y(3) = 6 - y(3)
  end do
  !$omp end parallel do
end
```
it generates
```
SUBROUTINE f00 (x, y)
 INTEGER x(:), y(:)
 INTEGER i
  reduction_temp_x(2)=x(2_8)
  reduction_temp_y(3)=y(3_8)
!$OMP PARALLEL DO REDUCTION(+: reduction_temp_x(2),reduction_temp_y(3))
 DO i=1_4,200_4
   x(2_8)=x(2_8)+i    ! unmodified
   reduction_temp_y(3)=6_4-reduction_temp_y(3)
 END DO
!$OMP END PARALLEL DO
  y(3_8)=reduction_temp_y(3)
END SUBROUTINE
```
leaving `x(2) = x(2) + i` unmodified.  Is that a known limitation?

Also, if you modify the testcase a bit:
```
subroutine f00(x, y, j)
  integer :: x(:), y(:)
  integer :: i, j
  !$omp parallel do reduction(+: x(2), y(j))
  do i = 1, 200
    x(2) = x(2) + i
    y(j) = 6 - y(j)
  end do
  !$omp end parallel do
end
```

you get incorrect output:
```
SUBROUTINE f00 (x, y, j)
 INTEGER x(:), y(:)
 INTEGER i, j
  reduction_temp_x(2)=x(2_8)
  reduction_temp_y(2)=y(int(j,kind=8))
!$OMP PARALLEL DO REDUCTION(+: reduction_temp_x(2),reduction_temp_y(2))
 DO i=1_4,200_4
   reduction_temp_y(2)=reduction_temp_y(2)+i
   reduction_temp_y(2)=6_4-y(int(j,kind=8))
 END DO
!$OMP END PARALLEL DO
  y(int(j,kind=8))=reduction_temp_y(2)
END SUBROUTINE
```

Stylie777 committed:
Fix issues and respond to comments

There were some corner case issues that were identified when using
multiple array elements in a reduction, along with using elements
that using a variable to identify which element to use, rather than
an int literal constant. Both of these have been fixed and tests
added to ensure future support.

Stylie777 committed:
formatting

Stylie777 reviewed (commented):
@kparzysz The issues you have raised were indeed issues that needed addressing. I have improved the implementation to address those issues and added tests to cover these too.

kparzysz commented on flang/lib/Semantics/rewrite-parse-tree.cpp:
```diff
@@ -492,10 +573,497 @@ void RewriteMutator::Post(parser::WriteStmt &x) {
+            [&](auto &) {}),
+        expr.u);
+  };
+  for (parser::SectionSubscript &subscript : sectionSubscript) {
```
The temp needs to store all subscripts.  Consider this:
```
subroutine f00(x, y, j)
  integer :: x(:), y(:,:)
  integer :: i, j
  if (j /= 2) then
    !$omp parallel do reduction(+: y(j, 2), y(2, 2))
    do i = 1, 200
      y(j, 2) = i - y(j, 2)
      y(2, 2) = y(2, 2) + 1
    end do
    !$omp end parallel do
  end if
end
```
We have two distinct array elements in the reduction, but the transformation only updates one of them:
```
SUBROUTINE f00 (x, y, j)
 INTEGER x(:), y(:,:)
 INTEGER i, j
 IF (j/=2_4) THEN
   reduction_temp_y(2)=y(int(j,kind=8),2_8)
   reduction_temp_y(2)=y(2_8,2_8)
!$OMP PARALLEL DO REDUCTION(+: reduction_temp_y(2),reduction_temp_y(2))
  DO i=1_4,200_4
    reduction_temp_y(2)=i-reduction_temp_y(2)
    reduction_temp_y(2)=reduction_temp_y(2)+1_4
  END DO
!$OMP END PARALLEL DO
   y(int(j,kind=8),2_8)=reduction_temp_y(2)
   y(int(j,kind=8),2_8)=reduction_temp_y(2)
 END IF
END SUBROUTINE
```

Stylie777 commented:
Closing this in favour of https://github.com/llvm/llvm-project/pull/196094
